modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageManagerImpl.java

```diff
@@ -311,75 +316,37 @@ public void addElectionListener(ElectionListener listener) {
         electionListeners.add(listener);
     }
 
-    private CompletableFuture<Long> recover(MetaStorageService service) {
-        if (!busyLock.enterBusy()) {
-            return failedFuture(new NodeStoppingException());
-        }
+    private CompletableFuture<?> updateTargetRevisionsForRecovery(MetaStorageService service) {
+        return inBusyLockAsync(busyLock, () -> {
+            service.currentRevisions()
+                    .thenAccept(targetRevisions -> {
+                        assert targetRevisions != null;
 
-        try {
-            service.currentRevision().whenComplete((targetRevision, throwable) -> {
-                if (throwable != null) {
-                    recoveryFinishedFuture.completeExceptionally(throwable);
+                        LOG.info("Performing MetaStorage recovery: [from={}, to={}]", storage.revisions(), targetRevisions);
 
-                    return;
-                }
+                        recoveryRevisionsListener.setTargetRevisions(targetRevisions.toRevisions());
+                    }).whenComplete((res, throwable) -> {
+                        if (throwable != null) {
+                            recoveryFinishedFuture.completeExceptionally(throwable);
+                        }
+                    });
 
-                LOG.info("Performing MetaStorage recovery from revision {} to {}", storage.revision(), targetRevision);
+            return recoveryFinishedFuture.whenComplete((revisions, throwable) -> {
+                if (throwable != null) {
+                    LOG.info("Recovery failed", throwable);
+                } else {
+                    long recoveryRevision = revisions.revision();
 
-                assert targetRevision != null;
+                    appliedRevision = recoveryRevision;
 
-                listenForRecovery(targetRevision);
-            }).whenComplete((res, ex) -> {
-                if (ex != null) {
-                    LOG.info("Recovery failed", ex);
+                    if (recoveryRevision > 0) {
+                        clusterTime.updateSafeTime(storage.timestampByRevision(recoveryRevision));
+                    }
 
-                    recoveryFinishedFuture.completeExceptionally(ex);
+                    LOG.info("Finished MetaStorage recovery");
```

Comment:
Please add a try/catch around this block; otherwise, if something is thrown from it, we'll never know it, just the future will never be completed, and it will be difficult to see what happened

Reply:
Fix it a little differently.